modules/core/src/test/java/org/apache/ignite/internal/util/tostring/TransactionSensitiveDataTest.java

```diff
@@ -0,0 +1,320 @@
+import static org.apache.ignite.transactions.TransactionIsolation.REPEATABLE_READ;
+
+/**
+ * Class for checking sensitive data when outputting transactions to the log.
+ */
+public class TransactionSensitiveDataTest extends GridCommonAbstractTest {
+    /** Class rule. */
+    @ClassRule public static final TestRule classRule = new SystemPropertiesRule();
+
+    /** Listener log messages. */
+    private static ListeningTestLogger testLog;
+
+    /** Network timeout. */
+    private static final long NETWORK_TIMEOUT = 500;
+
+    /** Node count. */
+    private static final int NODE_COUNT = 2;
+
+    /** Create a client node. */
+    private boolean client;
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        setFieldValue(GridNearTxPrepareFutureAdapter.class, "log", null);
+        ((AtomicReference)getFieldValue(GridNearTxPrepareFutureAdapter.class, "logRef")).set(null);
+
+        testLog = new ListeningTestLogger(false, log);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        testLog.clearListeners();
+
+        stopAllGrids();
+
+        clearGridToStringClassCache();
+
+        super.afterTest();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setConsistentId(igniteInstanceName)
+            .setGridLogger(testLog)
+            .setClientMode(client)
+            .setCommunicationSpi(new TestRecordingCommunicationSpi())
+            .setCacheConfiguration(
+                new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+                    .setAtomicityMode(TRANSACTIONAL)
+                    .setBackups(3)
+                    .setAffinity(new RendezvousAffinityFunction(false, 10))
+            )
+            .setNetworkTimeout(NETWORK_TIMEOUT);
+    }
+
+    /**
+     * Test for checking the absence of sensitive data in log during an
+     * exchange while an active transaction is running.
+     *
+     * @throws Exception If failed.
+     */
+    @WithSystemProperty(key = IGNITE_TO_STRING_INCLUDE_SENSITIVE, value = "false")
+    @Test
+    public void testHideSensitiveDataDuringExchange() throws Exception {
+        checkSensitiveDataDuringExchange((s1, s2) -> assertNotContains(log, s1, Person.class.getSimpleName()));
+    }
+
+    /**
+     * Test for checking the presence of sensitive data in log during an
+     * exchange while an active transaction is running.
+     *
+     * @throws Exception If failed.
+     */
+    @WithSystemProperty(key = IGNITE_TO_STRING_INCLUDE_SENSITIVE, value = "true")
+    @Test
+    public void testShowSensitiveDataDuringExchange() throws Exception {
+        checkSensitiveDataDuringExchange((s1, s2) -> assertContains(log, s1, s2));
```

Comment:
what do you check here ? existence already parsed string ?

Reply:
I verify that "BinaryObject <Person> #toString ()" is present in the log message. 
Made it so as not to duplicate the code.